### PR TITLE
*: add a API to clear tombstone store 

### PR DESCRIPTION
**Pull request by bradyjoestar**
### What problem does this PR solve? 
#1468 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
 - No code

Code changes

 - Has configuration change
 - Has HTTP API interfaces change
 - Has persistent data change

Side effects

 - Possible performance regression
 - Increased code complexity
 - Breaking backward compatibility

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation
 - Need to update the `tidb-ansible` repository
 - Need to be included in the release notes

### CODE REVIEW

**Comment by nolouch** [comment]
@bradyjoestar Can you add a command like
 ```
store clear [id]
``` 
in `tools/pd-ctl`.
and add some test in `tests/cmd/pdctl_test.go`

**Comment by bradyjoestar** [comment]
@nolunch sure.
I am a newbie so it may takes some time to read the testcode.

File: server/cluster.go [hunk 1]
```diff
@@ -557,6 +557,13 @@ func (c *RaftCluster) checkStores() {
 	}
 }
 
+// GetStores gets stores from cluster.
```
Comment (rleungx) [inline server/cluster.go]: It seems this comment is not for this function?

File: server/core/store.go [hunk 2]
```diff
@@ -554,6 +554,15 @@ func (s *StoresInfo) GetMetaStores() []*metapb.Store {
 	return stores
 }
 
+// RemoveTombStoneStores remove all the record of tombstone Store.
```
Comment (rleungx) [inline server/core/store.go]: ditto

Comment (bradyjoestar) [reply server/core/store.go]: Fixed!Thanks.

File: server/cluster.go (hunk 1 above)
Comment (bradyjoestar) [reply server/cluster.go]: Yeah, :grin:

File: server/core/store.go [hunk 3]
```diff
@@ -554,6 +554,15 @@ func (s *StoresInfo) GetMetaStores() []*metapb.Store {
 	return stores
 }
 
+// RemoveTombStoneRecords remove all the record of tombstone Records.
+func (s *StoresInfo) RemoveTombStoneRecords() {
+	for storeID, store := range s.stores {
+		if store.IsTombstone() {
+			delete(s.stores, storeID)
```
Comment (nolouch) [inline server/core/store.go]: Only remove in the memory, restarting PD will appear again.

Comment (bradyjoestar) [reply server/core/store.go]: Fixed , Thanks !

File: server/api/store.go [hunk 4]
```diff
@@ -186,6 +186,18 @@ func (h *storeHandler) Delete(w http.ResponseWriter, r *http.Request) {
...
 }
 
+func (h *storeHandler) RemoveTombStone(w http.ResponseWriter, r *http.Request) {
+	cluster := h.svr.GetRaftCluster()
+	if cluster == nil {
+		errorResp(h.rd, w, errcode.NewInternalErr(server.ErrNotBootstrapped))
+		return
+	}
+
+	cluster.RemoveTombStoneRecords()
```
Comment (nolouch) [inline server/api/store.go]: should handle the error.

File: server/cluster.go [hunk 5]
```diff
@@ -557,6 +557,25 @@ func (c *RaftCluster) checkStores() {
...
+	cluster := c.cachedCluster
+
+	for _, store := range cluster.GetStores() {
+		if store.IsTombstone() {
+			// the store has already been tombstone
+			err := cluster.deleteStore(store)
+			if err != nil {
+				return err
+			}
+		}
```
Comment (nolouch) [inline server/cluster.go]: Better to add a log if delete successfully.

File: server/api/router.go [hunk 6]
```diff
@@ -64,6 +64,7 @@ func createRouter(prefix string, svr *server.Server) *mux.Router {
 	storeHandler := newStoreHandler(svr, rd)
 	router.HandleFunc("/api/v1/store/{id}", storeHandler.Get).Methods("GET")
 	router.HandleFunc("/api/v1/store/{id}", storeHandler.Delete).Methods("DELETE")
+	router.HandleFunc("/api/v1/store/remove-tombstone", storeHandler.RemoveTombStone).Methods("DELETE")
```
Comment (rleungx) [inline server/api/router.go]: I think it's better to put it into `stores` instead of `store`?

File: server/cluster.go [hunk 7]
```diff
@@ -557,6 +557,28 @@ func (c *RaftCluster) checkStores() {
 	}
 }
 
+// RemoveTombStoneRecords remove the tombStone Records.
```
Comment (rleungx) [inline server/cluster.go]: s/remove/removes/g

File: tools/pd-ctl/pdctl/command/store_command.go [hunk 8]
```diff
@@ -70,6 +70,15 @@ func NewSetStoreWeightCommand() *cobra.Command {
 	}
 }
 
+// NewRemoveTombStoneCommandFunc returns a weight subcommand of storeCmd.
```
Comment (rleungx) [inline tools/pd-ctl/pdctl/command/store_command.go]: weight?

File: server/core/kv.go [hunk 9]
```diff
@@ -114,6 +114,11 @@ func (kv *KV) SaveStore(store *metapb.Store) error {
 	return saveProto(kv.KVBase, kv.storePath(store.GetId()), store)
 }
 
+// DeleteStore delete one store from KV.
```
Comment (rleungx) [inline server/core/kv.go]: s/delete/deletes/g

File: server/cluster.go (hunk 7 above)
Comment (bradyjoestar) [reply server/cluster.go]: could you please tell me what the meaning of "s/remove/removes/g"? little confused. Thanks you!

Comment (bradyjoestar) [reply server/cluster.go]: @rleungx

File: server/api/router.go (hunk 6 above)
Comment (bradyjoestar) [reply server/api/router.go]: Fine, I'll move it to `storesHandler` tonight. :ok_hand:

File: server/cluster.go (hunk 7 above)
Comment (rleungx) [reply server/cluster.go]: The verbs, which use a function name as the first person, should use the third-person-singular. e.g.
```suggestion
// RemoveTombStoneRecords removes the tombStone Records.
```
They should be changed all in this PR.

File: server/core/kv.go (hunk 9 above)
Comment (bradyjoestar) [reply server/core/kv.go]: Next time I will pay attention to it. :grin:

File: server/cluster.go (hunk 5 above)
Comment (nolouch) [reply server/cluster.go]: We want to know the deleted store through the log, so we also need to print the log if it is successfully deleted.

Comment (bradyjoestar) [reply server/cluster.go]: Got it. :ok_hand:

**Pull request merged**